Compose dev

jbaptperez opened the pull request:
**IMPORTANT: All Pull Requests should be connected to an issue, if you don't
have an issue, please start by creating an issue and link it to the PR.**

Please provide enough information so that others can review your pull request:

<!-- You can skip this if you're fixing a typo or adding a minor bug fix -->

+ What existing problem does this PR solve?
  - Issue #3225 (Adapt the development Compose set-up to allow restricted company contexts).
+ What new feature is being introduced with this PR?
  - Add to the development set-up the possibility to customise image builds (among other) using a `.env` file,
  - Update of the Compose file syntax to the Compose specification,
  - Other minor enhancements related to the development Compose set-up
+ Overview of changes to existing functions if required.
  - The current behavior remain the same, but the enhancement of the development set-up leads to different paths to the Dockerfiles and simplified commands to manage the development container.

<!-- Example: This PR adds a function to do X, which adds the ability to do Y.
-->

**Checks**

- [x] All tests succeed.
- [ ] Unit tests added.
- [ ] e2e tests added.
- [x] Documentation updated.

**Closing issues**

Closes #3225.

## Changes

### .gitattributes and .dockerignore

Adds .gitattributes file to handle line endings, in particular when developing on Windows and building Linux images.
Adds .dockerignore file to limit the context transferred when building Docker images.

### New Compose set-up for development

- Located at `contrib/docker/dev`.
- **Does NOT replace the `docker/dev` set-up** for now.
- Uses a **latest Compose syntax**.
- Relies only on Docker Compose CLI, not both Docker Compose and Docker CLIs anymore.
- **Allows image builds in enterprise context**: Optional dependency source proxies can be set for Ubuntu, Python and Node.
- Uses **optimized multi-stage builds** to locally build images.
- Centralizes variables in an non-versioned `.env` file, with a `.env.template` model file.
- Uses a structured file hierarchy for Compose services.
- Uses per service environment files.
- **Uses only named volumes** for speed, portability and to store all data that are not configuration or source files (use `docker compose down -v` to reset).
- **Splits the main service** (container) into one per component (setup, celery-worker, gunicorn, frontend).
- Makes the frontend service compatible with the 3 frontends implementations (switch via an environment variable).
- **Dramatically optimizes the setup stage (dedicated service)** (shallow clones/fetches the Sigma repository, calls `tsctl` once to register all rules at once).
- Uses a single entry point for each Docker service.
- **Makes each service run on PID 1** to properly receive Docker signals.**
- **Optimizes service startup order and leverages health checks** to make service wait for each other (PostgreSQL, Redis, Gunicorn). OpenSearch has also a health check but the healthy state is not strictly required for development purposes.
- **Adds a dedicated `README.md`** to quickly get started.

### Limits

I used a copy of the `timesketch.conf` file to properly set variables required for a local setup.
This copy must be updated according to the modification mades to the `data/timesketch.conf` file.

However, if you decides to replace the current Compose development setup with the new (contrib) one, this can be solved the following way:

- Move `contrib/docker/dev` to `docker/dev`,
- Adapt content (relative paths and README.md commands),
- Declare an environment variable for all settings that can be customized in the (Compose) `timesketch.conf` file,
- Adapt and test using Compose,
- Move the Compose `timesketch.conf` and replace the `data/` one,
- Adapt the Compose references to the new location (bind mounts),
- Find references that use this file (development or contribution scripts) and adapt them: just declare and use environment variables.

jbaptperez committed:
Update .dockerignore

Adds .gitignore files.
Dramatically improves an image build in a development context.

jbaptperez committed:
Add a .gitattribute file

Makes the repository handle file line endings.
This helps to make it cross-platform, asserting some files are Unix-ended.

jbaptperez commented:
As I had time to update the PR, I also optimized some things, like using a Python virtual environment in the Docker image.
There is a lot of another improvements to be done, but maybe in another PR?

I thought about:

- Adding an environment variable to allow a custom place for the configuration files (with a default value to be compatible), this is necessary for the next steps,
- Put every component into its own container,
- Using Compose profiles to start everything or only a set of services,
- Allow, with port mapping to localhost, using our IDE to start a process in true debug mode (_gunicorn_ or _flask_, _celery_),
- Use Compose development features for an optimized synchronization with the current source code.

jbaptperez commented:
@jkppr, actually, the apparent change is not so big: Everything run into a single container with `docker compose exec` commands instead of `docker exec` ones.

I paid special attention to keeping default settings so that the behaviour without changes is the same as before.

A big change reported by Git is the deletion of the recently added `yarn.lock` file as it freezes a package source repository URL, which is indeed an issue for the PR.

The rest is close to the previous behaviour, with a single real difference: I duplicated the `timesketch.conf` file so that the development Compose set-up have its own; This simplifies the configuration from a single `.env` file but this is a kind of duplication, which can be discussed.

However, I understand your message and I am open to suggestions.

jaegeral commented:
Is there a way to have those files maybe in the contrib folder so we can keep the way it is currently done and a kind of forked thing in `contrib`

jbaptperez committed:
Add a new Compose for development

Compose standard changes:
- Removes the deprecated "version" field,
- Adds a toplevel "name" field (prefix of container names nad network),
- Adds a toplevel "network" field, with a common "timesketch-dev"
  network,
- Removes container names (depends on and toplevel name and service
  names),
- Do not bind to the 127.0.0.1 interface only (0.0.0.0),
- Removes useless "links" (common network),
- Refactors environment variables not to use a YAML array,
- Removes "restart" fields to detect undesired crashes in development,
- Binds ports of other services to the host (opensearch, redis).

General changes:
- Allows Docker image builds in a restricted company context (limited
  access to remote Ubuntu, Python or Node repositories) using variables,
- Centralizes variables in a .env file (not versioned),
- Adds a .env.template file as .env template with predefined variables,
- Use a distinct directory for every service dependencies,
- Use named volumes for portability and to avoid auto-creation of
  anonymous ones (PostgreSQL, Redis and Prometheus declare volumes in
  their Dockerfile; this leads to anonymous volume creations if they
  are not declared in Compose),
- Uses a per-service environment file,
- Uses local images that include their healthcheck,
- Rename some service names,
- Simplifies how development configuration files are transferred to
  Timesketch,
- Simplifies manipulation of containers using Compose CLI instead of
  the Docker one,
- Simplify and optimizes the Timesketch entrypoint,
- Updates the Bash scripts to start frontend-ng,
- Updates related documentation.

jaegeral commented on .gitignore:
```diff
@@ -48,4 +47,12 @@ vagrant/*.log
 
 # Exclude .venv folder
 .venv/
+
```
Please make a seperate PR if you intend to change things in the .gitignore (and other . files)

jbaptperez replied:
Done. See [.gitignore enhancement PR](https://github.com/google/timesketch/pull/3423).

jaegeral commented on docker/dev/README.md:
````diff
@@ -7,41 +7,46 @@ NOTE: It is not recommended to try to run on a system with less than 8 GB of RAM
 ### Start a developer version of docker containers in this directory
 
-```
+```bash
````
Those changes in the file looks good, consider branching it out to a seperate PR (changes in docs can be merged much quicker in most cases)

jbaptperez replied:
Done. See [Documentation fixes PR](https://github.com/google/timesketch/pull/3424).

jaegeral commented on docs/developers/frontend-development.md:
```diff
@@ -4,7 +4,7 @@ hide:
 ### Frontend (old) development dependencies
 
-When developing the frontend you use the VueJS frontend server. Changes will be picked up automatically
+When developing the frontend you use the Vue.js frontend server. Changes will be picked up automatically
```
same as the other docs/ file, happy to merge that quickly in a seperate PR.

jbaptperez replied:
Done. See [Documentation fixes PR](https://github.com/google/timesketch/pull/3424).

jaegeral reviewed (changes requested):
Some comments to address, I will try to have a look at the rest of the PR potentially this week.

jbaptperez commented on docker/dev/build/sigma_rules.txt:
```diff
@@ -1,7 +1,6 @@
-/usr/local/src/sigma/rules/application/spring/appframework_spring_exceptions.yml
 /usr/local/src/sigma/rules/application/sql/app_sqlinjection_errors.yml
-/usr/local/src/sigma/rules/other/godmode_sigma_rule.yml
+/usr/local/src/sigma/other/godmode_sigma_rule.yml
```
Well, the Sigma rules repository structure changed.
Adapting the rule just helps to make the things work again.
The entry point shell script checks the file exists and warns in it does not.
This is not a breaking point so this can also be put to a different PR if you prefer.

jbaptperez commented:
@jaegeral, for now I keep changes in this PR untill other individual PRs are merged.
Afterwards, I'll rebase everything and already merged commits will be dropped.

jbaptperez commented on contrib/docker/dev/notebook/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+FROM us-docker.pkg.dev/osdfir-registry/picatrix/picatrix:latest
```
I could not find a proper tag.

jbaptperez commented on contrib/docker/dev/timesketch/docker-entrypoint.sh:
```diff
@@ -0,0 +1,54 @@
+  exit 0
+}
+
+. "${HOME}/.virtualenvs/app/bin/activate"
```
Done.

jbaptperez commented on contrib/docker/dev/README.md:
```diff
@@ -0,0 +1,251 @@
+
+---
+
+## TODO Formating README.md
```
To be done in the future.

jbaptperez committed:
Use a single Compose service per development component

Dockerfile changes:
- Switches to optimized multistage build,
- Switch to rootless images (user name, UID and GID are configurable),
- Use dedicated entrypoints per images,
- Add a health check for the Gunicorn image,
- Frontend compatible with all 3 versions at build time.

Compose changes:
- Splits services into multiple ones,
- Use a dedicated setup service for data initialization,
- Dramatically optimizes the initial setup.

README.md changes:
- Rewrites everything, clearer.

jbaptperez committed:
Make backend proxy configurable for frontend development

Required to isolate development processes into their own Docker
container.

jaegeral committed:
Merge branch 'master' into compose-dev

sasirven commented:
Hi any news about this Feature @jaegeral?

jaegeral commented:
Hey @sasirven, I know I indicated in https://github.com/google/timesketch/pull/3226#issuecomment-2750470699 if it would be an option to put it in contrib, but now looking at it how it plaid out, I am concerned that this will create a nightmare to maintain. Every change we make to e.g. the timesketch.conf would have to be done in two places, we added OpenTelemetry to the project in https://github.com/google/timesketch/pull/3805. The once `the apparent change is not so big` became a 1000 LOC addition with a lot of stuff being duplicate to other places in the repo.

Is there maybe a more robust way to get what you want to achieve? E.g. a new folder in contrib with a README.md with the problem description and a shell script that when executed copies all the files needed in place?

It also goes way beyond the initial idea to this: https://github.com/google/timesketch/issues/3225 looking at this lis:

```
Located at contrib/docker/dev.
Does NOT replace the docker/dev set-up for now.
Uses a latest Compose syntax.
Relies only on Docker Compose CLI, not both Docker Compose and Docker CLIs anymore.
Allows image builds in enterprise context: Optional dependency source proxies can be set for Ubuntu, Python and Node.
Uses optimized multi-stage builds to locally build images.
Centralizes variables in an non-versioned .env file, with a .env.template model file.
Uses a structured file hierarchy for Compose services.
Uses per service environment files.
Uses only named volumes for speed, portability and to store all data that are not configuration or source files (use docker compose down -v to reset).
Splits the main service (container) into one per component (setup, celery-worker, gunicorn, frontend).
Makes the frontend service compatible with the 3 frontends implementations (switch via an environment variable).
Dramatically optimizes the setup stage (dedicated service) (shallow clones/fetches the Sigma repository, calls tsctl once to register all rules at once).
Uses a single entry point for each Docker service.
Makes each service run on PID 1 to properly receive Docker signals.**
Optimizes service startup order and leverages health checks to make service wait for each other (PostgreSQL, Redis, Gunicorn). OpenSearch has also a health check but the healthy state is not strictly required for development purposes.
Adds a dedicated README.md to quickly get started.
```

Which in itself might be better suited smaller PRs (or issues in the first place to discuss the problem to be solved). E.g.: `Uses only named volumes for speed, portability and to store all data that are not configuration or source files (use docker compose down -v to reset).` how does that contribute to the initial problem? If it is adding value, it could be in the main docker/dev file, or else we are creating a fork of the docker/dev file that, as stated above, becomes a nightmare to maintain and troubleshoot, as we would have to ask everyone who has problems, which setup they are using.

sasirven commented:
Hey @jaegeral, sorry for the delay.

I agree that duplication would be a nightmare for you to maintain.
After looking at this in more detail, here’s my proposed approach:

- Provide a small script or clear instructions in the README explaining that users should copy the data folder locally in the `docker/dev` and override only the files they need (for example, timesketch.conf to set personal values). We'll need to update the compose to properly mount this custom folder instead of the root data one to work.
- Add an environment variable to properly mount the CA certificates so it works in a company environment.
- Include a .env.example file with all customizable values (e.g., pip and node PPA, CA certificates, etc.).
- For the release Docker setup, add a variable to select the repository for the Timesketch image, in case a company needs to override it.

This should make it much easier to run the project locally in a corporate environment withtout the maintenance nightmare.

I also noticed an issue with OIDC (this should be tracked in a separate issue): it currently requires an https scheme for the redirect_url in google_auth.py. This should be configurable via an environment variable (or similar) to support local OIDC redirects.
